src/optimizer/build_probe_side_optimizer.cpp

```diff
@@ -64,6 +64,37 @@ static inline idx_t ComputeOverlappingBindings(const vector<ColumnBinding> &hays
 	return result;
 }
 
+BuildSize BuildProbeSideOptimizer::GetBuildSizes(LogicalOperator &op) {
+	BuildSize ret;
+	switch (op.type) {
+	case LogicalOperatorType::LOGICAL_COMPARISON_JOIN:
+	case LogicalOperatorType::LOGICAL_CROSS_PRODUCT:
+	case LogicalOperatorType::LOGICAL_ANY_JOIN:
+	case LogicalOperatorType::LOGICAL_DELIM_JOIN: {
+
+		auto &left_child = op.children[0];
+		auto &right_child = op.children[1];
+
+		// resolve operator types to determine how big the build side is going to be
+		op.ResolveOperatorTypes();
+		auto left_tuple_layout = TupleDataLayout();
+		left_tuple_layout.Initialize(left_child->types);
+
+		auto right_tuple_layout = TupleDataLayout();
+		right_tuple_layout.Initialize(right_child->types);
+
+		// Don't multiply by cardinalities, the only important metric is the size of the row
+		// in the hash table
+		ret.left_side = left_tuple_layout.GetRowWidth() + COLUMN_COUNT_PENALTY * left_child->types.size();
```

Comment:
This then also needs to use `left_types.size()`.

Also, I don't think _adding_ the penalty to the row width is the right way to go. I think it should be some kind of _multiplication_. Perhaps we can multiply the row width by `(1 + 0.1 * left_types.size())`?

Let's compare 2 `HUGEINT` columns vs. 15 `TINYINT`. I haven't done any experiments, but I would assume that building on 2 hugeint columns is cheaper than building on 15 tinyints.

Hugeints cost:
```c++
idx_t width = 1 + 16*2 + 8; // 41 validitymask, hugeints, hash (3 columns total)
idx_t add_cost = width + 1.1 * 3; // 44.3
idx_t mult_cost = width * (1 + 0.1 * 3); // 53.3
```

Tinyints cost:
```c++
idx_t width = 2 + 1*15 + 8; // 25 validitymask, tinyints, hash (16 columns total)
idx_t add_cost = width + 1.1 * 16; // 42.6
idx_t mult_cost = width * (1 + 0.1 * 16); // 65.0
```

Here, `add_cost` is your current approach, and `mult_cost` is what I'm proposing.

As we can see, with your current approach, the cost of building on two hugeints is higher than the cost of building on 15 tinyints. I would expect that building on 15 tinyints is more costly, though. With the formula I'm proposing, the cost of building on the tinyints is higher.